lib/gc/Dialect/Linalgx/Utils.cpp

```diff
@@ -0,0 +1,451 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Parser/Parser.h"
+
+#include "gc/Dialect/Linalgx/Utils.h"
+
+namespace mlir {
+namespace linalgx {
+
+//****************************************************************************//
+//                      Packed Matmul Structs                                 //
+//****************************************************************************//
+
+struct BatchDimMap {
+public:
+  BatchDimMap() = default;
+  BatchDimMap(ArrayRef<int64_t> batchARef, ArrayRef<int64_t> batchBRef,
+              ArrayRef<int64_t> batchCRef)
+      : batchA(batchARef), batchB(batchBRef), batchC(batchCRef) {}
+  /// Get original arrays
+  ArrayRef<int64_t> getBatchA() const { return ArrayRef<int64_t>(batchA); }
+  ArrayRef<int64_t> getBatchB() const { return ArrayRef<int64_t>(batchB); }
+  ArrayRef<int64_t> getBatchC() const { return ArrayRef<int64_t>(batchC); }
+  /// Get size of batch dims
+  int64_t getBatchNum() const { return batchA.size(); }
+  /// Get attr for batch info
+  bool isBatchEmpty() const {
+    return batchA.empty() && batchB.empty() && batchC.empty();
+  }
+  bool isBatchMatmul() const {
+    return !isBatchEmpty() && batchEqualAB() && batchEqualAC();
+  }
+  bool isBatchReduce() const {
+    return !isBatchEmpty() && batchEqualAB() && batchC.empty();
+  }
+
+private:
+  bool batchEqualAB() const {
+    return std::equal(batchA.begin(), batchA.end(), batchB.begin());
+  }
+  bool batchEqualAC() const {
+    return std::equal(batchA.begin(), batchA.end(), batchC.begin());
+  }
+  SmallVector<int64_t> batchA;
+  SmallVector<int64_t> batchB;
+  SmallVector<int64_t> batchC;
+};
+
+struct PackingMap {
+public:
+  PackingMap(ArrayRef<int64_t> firstRef, ArrayRef<int64_t> secondRef)
+      : first(firstRef), second(secondRef) {}
+  /// Get original arrays
+  ArrayRef<int64_t> getFirst() const { return ArrayRef<int64_t>(first); }
+  ArrayRef<int64_t> getSecond() const { return ArrayRef<int64_t>(second); }
+  /// SrcDims.size() == 1; DstDims.size() >= 1
+  ArrayRef<int64_t> getPackingSrcDims() const {
+    return getPackingSrcIndex() == 0 ? getFirst() : getSecond();
+  }
+  ArrayRef<int64_t> getPackingDstDims() const {
+    return getPackingDstIndex() == 0 ? getFirst() : getSecond();
+  }
+  /// Index first is 0; Index second is 1
+  unsigned getPackingSrcIndex() const { return getFirst().size() == 1 ? 0 : 1; }
+  unsigned getPackingDstIndex() const { return getFirst().size() == 1 ? 1 : 0; }
```

Comment:
In many cases both first and second can be with `size > 1`, then the result of the utility function might be wrong.

Reply:
> In many cases both first and second can be with `size > 1`, then the result of the utility function might be wrong.

Discussed with Longshen, there is no such possibility in matmul